pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/TaskMetricsEmitter.java

```diff
@@ -71,42 +78,105 @@ protected final void runTask(Properties periodicTaskProperties) {
-        int percent = total != 0 ? (accumulated.getWaiting() + accumulated.getRunning()) * 100 / total : 0;
+            taskTypeAccumulatedCount.getError());
+        int total = taskTypeAccumulatedCount.getTotal();
+        int percent = total != 0
+            ? (taskTypeAccumulatedCount.getWaiting() + taskTypeAccumulatedCount.getRunning()) * 100 / total : 0;
         _controllerMetrics.setValueOfGlobalGauge(ControllerGauge.PERCENT_MINION_SUBTASKS_IN_QUEUE, taskType, percent);
-        percent = total != 0 ? accumulated.getError() * 100 / total : 0;
+        percent = total != 0 ? taskTypeAccumulatedCount.getError() * 100 / total : 0;
         _controllerMetrics.setValueOfGlobalGauge(ControllerGauge.PERCENT_MINION_SUBTASKS_IN_ERROR, taskType, percent);
+
+        // Emit metrics for table taskType
+        tableAccumulatedCount.forEach((tableNameWithType, taskCount) -> {
+          _controllerMetrics.setOrUpdateTableGauge(tableNameWithType, taskType,
+              ControllerGauge.NUM_MINION_SUBTASKS_RUNNING, () -> (long) taskCount.getRunning());
+          _controllerMetrics.setOrUpdateTableGauge(tableNameWithType, taskType,
+              ControllerGauge.NUM_MINION_SUBTASKS_WAITING, taskCount.getWaiting());
+          _controllerMetrics.setOrUpdateTableGauge(tableNameWithType, taskType,
+              ControllerGauge.NUM_MINION_SUBTASKS_ERROR, taskCount.getError());
+          int tableTotal = taskCount.getTotal();
+          int tablePercent = tableTotal != 0 ? (taskCount.getWaiting() + taskCount.getRunning()) * 100 / tableTotal : 0;
+          _controllerMetrics.setOrUpdateTableGauge(tableNameWithType, taskType,
+              ControllerGauge.PERCENT_MINION_SUBTASKS_IN_QUEUE, tablePercent);
+          tablePercent = tableTotal != 0 ? taskCount.getError() * 100 / tableTotal : 0;
+          _controllerMetrics.setOrUpdateTableGauge(tableNameWithType, taskType,
+              ControllerGauge.PERCENT_MINION_SUBTASKS_IN_ERROR, tablePercent);
+        });
+
+
+        if (_preReportedTables.containsKey(taskType)) {
+          Set<String> tableNameWithTypeSet = _preReportedTables.get(taskType);
+          tableNameWithTypeSet.removeAll(tableAccumulatedCount.keySet());
+          removeTableTaskTypeMetrics(tableNameWithTypeSet, taskType);
+        }
+        if (!tableAccumulatedCount.isEmpty()) {
+          // need to make a copy of the set because we may want to chagne the set later
+          Set<String> tableNameWithTypeSet = new HashSet<>(tableAccumulatedCount.keySet());
+          _preReportedTables.put(taskType, tableNameWithTypeSet);
+        } else {
+          _preReportedTables.remove(taskType);
+        }
       } catch (Exception e) {
         LOGGER.error("Caught exception while getting metrics for task type {}", taskType, e);
       }
     }
 
+    // clean up metrics for task types that have already been removed
+    _preReportedTaskTypes.removeAll(taskTypes);
+    for (String taskType : _preReportedTaskTypes) {
```

Comment:
@zhtaoxiang We didn't clean up the metrics when the table is deleted?

Reply:
The previous code does not have the logic to clean up metrics, although it does not report table level metrics (only task type level metrics)